File: src/Java/pipeline-core/src/main/java/com/pipeline/core/PipelineObserver.java
package com.pipeline.core;

/** Optional observer for one canonical pipeline runner. Observer failures are ignored. */
public interface PipelineObserver {
  PipelineObserver NOOP = new PipelineObserver() {};

  default void onPipelineStarted(String pipelineName) {}

  default void onActionStarted(
      String pipelineName,
      StepPhase phase,
      int actionIndex,
      String actionName) {}

  default void onActionCompleted(
      String pipelineName,
      StepPhase phase,
      int actionIndex,
      String actionName,
      long elapsedNanos) {}

  default void onActionFailed(
      String pipelineName,
      StepPhase phase,
      int actionIndex,
      String actionName,
      Exception error,
      long elapsedNanos) {}

  default void onShortCircuited(
      String pipelineName,
      StepPhase phase,
      int actionIndex,
      String actionName) {}

  default void onPipelineCompleted(
      String pipelineName,
      boolean shortCircuited,
      int errorCount,
      long elapsedNanos) {}
}

File: src/CSharp/pipeline_services/core/PipelineObserver.cs
using System;

namespace PipelineServices.Core;

/// <summary>One optional observability seam over the canonical Pipeline runner.</summary>
public interface PipelineObserver
{
    void OnPipelineStarted(string pipelineName) { }

    void OnActionStarted(
        string pipelineName,
        StepPhase phase,
        int actionIndex,
        string actionName) { }

    void OnActionCompleted(
        string pipelineName,
        StepPhase phase,
        int actionIndex,
        string actionName,
        long elapsedNanos) { }

    void OnActionFailed(
        string pipelineName,
        StepPhase phase,
        int actionIndex,
        string actionName,
        Exception exception,
        long elapsedNanos) { }

    void OnShortCircuited(
        string pipelineName,
        StepPhase phase,
        int actionIndex,
        string actionName) { }

    void OnPipelineCompleted(
        string pipelineName,
        bool shortCircuited,
        int errorCount,
        long elapsedNanos) { }
}

internal sealed class NoopPipelineObserver : PipelineObserver
{
    internal static readonly NoopPipelineObserver Instance = new();
    private NoopPipelineObserver() { }
}
